tools/serverpod_cli/test/runner/better_command_runner/analytics_test.dart

```diff
@@ -0,0 +1,183 @@
+
+  MockCommand() {
+    argParser.addOption(
+      'name',
+      defaultsTo: 'serverpod',
+      allowed: <String>['serverpod'],
+    );
+  }
+}
+
+void main() {
+  late BetterCommandRunner runner;
+  group('Given runner with null onAnalyticsEvent callback', () {
+    var runner = BetterCommandRunner(
+      'test',
+      'this is a test cli',
+      onAnalyticsEvent: null,
+    );
+
+    test('when checking if analytics is enabled then false is returned.', () {
+      expect(runner.analyticsEnabled(), isFalse);
+    });
+
+    test('when checking available flags then analytics flag is not present.',
+        () {
+      expect(runner.argParser.options.keys, isNot(contains('analytics')));
+    });
+  });
+
+  group('Given runner with onAnalyticsEvent callback defined', () {
+    var runner = BetterCommandRunner(
+      'test',
+      'this is a test cli',
+      onAnalyticsEvent: (event) {},
+    );
+
+    test('when checking if analytics is enabled then true is returned.', () {
+      expect(runner.analyticsEnabled(), isTrue);
+    });
+
+    test('when checking available flags then analytics is defined.', () {
+      expect(runner.argParser.options.keys, contains('analytics'));
+    });
+  });
+
+  group('Given runner with analytics enabled', () {
+    List<String> events = [];
+    setUp(() {
+      runner = BetterCommandRunner(
+        'test',
+        'this is a test cli',
+        onAnalyticsEvent: (event) => events.add(event),
+      );
+      assert(runner.analyticsEnabled());
+    });
+
+    tearDown(() {
+      events = [];
+    });
+
+    test(
+        'when running command with no-analytics flag then analytics is disabled.',
+        () async {
+      var args = ['--no-${BetterCommandRunnerFlags.analytics}'];
+      await runner.run(args);
+
+      expect(runner.analyticsEnabled(), isFalse);
+    });
+
+    test('when running invalid command then "invalid" analytics event is sent.',
+        () async {
+      var args = ['this could be a command argument'];
+
+      try {
+        await runner.run(args);
+      } catch (_) {
+        // Ignore any exception
+      }
+
+      expect(events, hasLength(1));
+      expect(events.first, equals('invalid'));
+    });
+
+    test(
+        'when running with unknown command then "invalid" analytics event is sent.',
+        () async {
+      var args = ['--unknown-command'];
+
+      try {
+        await runner.run(args);
+      } catch (_) {
+        // Ignore any exception
+      }
+
+      expect(events, hasLength(1));
+      expect(events.first, equals('invalid'));
+    });
+
+    test('when running with no command then "help" analytics event is sent.',
+        () async {
+      await runner.run([]);
+
+      expect(events, hasLength(1));
+      expect(events.first, equals('help'));
+    });
+
+    test(
+        'when running with only registered flag then "help" analytics event is sent.',
+        () async {
+      await runner.run(['--${BetterCommandRunnerFlags.analytics}']);
```

Comment:
Suggestion (none blocking): A thought here, what do you think about hardcoding this value instead?

My thinking is that we are testing a known interface, and if you make changes to them I would probably expect some test to fail and would be worried we have no tests if that doesn't happen. It could potentially take more time just figuring out if we have tests or not than it would having to change the hardcoded value.

Reply:
How about we add a specific test that validates that this value is specified? Like a validation of all the flags we assume are there?

Reply:
That way we could give a better message when that test would fail.